src/hotspot/share/opto/printinlining.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "printinlining.hpp"
+#include "opto/callnode.hpp"
+#include "memory/allocation.hpp"
+#include "memory/resourceArea.hpp"
+
+InlinePrinter::InlinePrinter(Arena* arena, bool enabled) : _enabled(enabled), _root(new(arena) IPInlineSite(nullptr, arena, 0)) {}
+
+InlinePrinter::IPInlineAttempt::IPInlineAttempt(InliningResult result) : result(result) {}
+
+outputStream* InlinePrinter::record(ciMethod* callee, JVMState* state, InliningResult result, const char* msg) {
+  if (!_enabled) {
+    return &_nullStream;
+  }
+  IPInlineAttempt* attempt = locate(state, callee)->add(result);
+  if (msg != nullptr) {
+    attempt->msg.print("%s", msg);
+  }
+  return &attempt->msg; // IPInlineAttempts are heap allocated so this address is safe
+}
+
+void InlinePrinter::print_on(outputStream* tty) {
+  if (!_enabled) {
+    return;
+  }
+  _root->dump(tty, -1);
+}
+
+InlinePrinter::IPInlineSite* InlinePrinter::locate(JVMState* state, ciMethod* callee) {
+  if (state == nullptr) {
+    return _root;
+  }
+
+  return locate(state->caller(), nullptr)->at_bci(state->bci(), callee);
+}
+
```

Comment:
It looks like you are building a tree, just like InlineTree.  I wonder if it would make sense to unify them somehow in the future.

Reply:
I thought about this too. Should I open an RFE for this?

Reply:
Yes, please.  Thanks.